core/src/avm2/extract.rs

```diff
@@ -0,0 +1,173 @@
+use crate::avm2::events::Event;
+use crate::avm2::object::TObject as _;
+use crate::avm2::Activation;
+use crate::avm2::Object;
+use crate::avm2::Value;
+use crate::display_object::DisplayObject;
+use crate::string::AvmString;
+use std::cell::{Ref, RefMut};
+
+/// This trait is implemented for each type that can appear in the signature
+/// of a method annotated with `#[native]` (e.g. `bool`, `f64`, `Object`)
+pub trait ExtractFromVm<'a, 'gc>: Sized {
+    /// Attempts to extract `Self` from the provided `Value`.
+    /// If the extraction cannot be performed (or would require a coercion),
+    /// then this should return `None`.
+    ///
+    /// The provided `activation` should only be used for debugging,
+    /// or calling a method that requires a `MutationContext`.
+    /// Any coercions (e.g. `coerce_to_string`) should have already been
+    /// performed by the time this method is called.
+    fn extract_from(val: &'a Value<'gc>, activation: &mut Activation<'_, 'gc, '_>) -> Option<Self>;
+}
+
+/// Allows writing `arg: DisplayObject<'gc>` in a `#[native]` method
+impl<'a, 'gc> ExtractFromVm<'a, 'gc> for DisplayObject<'gc> {
```

Comment:
Personally I think we should be implementing the trait in the file of the type, rather than the file of the trait.

Reply:
In this case, I think it's nice that it's in one file along with the primitives like `bool`. I also feel like this code will continue to change as we improve, so it's nice that it's in one place.